Docs: use parameter list for sqlite3.Cursor.execute* (#101782)
python/cpython: Doc/library/sqlite3.rst

   .. method:: execute(sql, parameters=(), /)

      Execute SQL a single SQL statement,
      optionally binding Python values using
      :ref:`placeholders <sqlite3-placeholders>`.

>       :param unicode sql:

Reviewer: (Possibly a dumb question; apologies in advance!) — Why "unicode" and not "str"?

Author: Lack of coffee, I guess.

         A single SQL statement.

      :param parameters:
         Python values to bind to placeholders in *sql*.
         A :class:`!dict` if named placeholders are used.
         A :term:`!sequence` if unnamed placeholders are used.
         See :ref:`sqlite3-placeholders`.

>       :type parameters: :class:`dict` | :term:`sequence`

Reviewer: It's weird to me that it'll accept any kind of sequence, but only one kind of mapping (`dict`). (No change requested.)

Author: Yes. Sequence or mapping would have been better. Perhaps future enhancement, but I remember that possibly being a breaking change.

      :raises ProgrammingError:

>          If *sql* contains more than one SQL statement.

Reviewer: > Also, I'm in doubt of how to use the :raises Sphinx param here; execute*() can raise a heck of a lot of exceptions; should we write all of them down here, or just a subset? Perhaps just point to the reference for sqlite3 exceptions?

Yeah, I feel like documenting exceptions that are frequently raised is Good, but if we leave it like this, we'll inevitably end up with "bug reports" saying "The sqlite documentation promises that only ProgrammingError will be raised, but I got a TypeError!!!1!1" or something along those lines.

Author: True that. I'll save that for a follow-up PR, though. It'll take some time.

Reviewer: Hmm, I feel like the current wording makes clear that `ProgrammingError` is not the only exception that _could_ be raised, whereas using the `:raises:` Sphinx param could imply that the docs are promising that it's the only exception that will ever be raised. So maybe we should not use `:raises:` here, and leave the language on `ProgrammingError` as it is (at least for now)?

Author: Yeah, that's a good idea.

Author: OTOH, I can just queue a PR that covers all exceptions later today (or tomorrow), so lets just keep it as it is on the PR.

      and there is no open transaction,
      a transaction is implicitly opened before executing *sql*.

      Use :meth:`executescript` to execute multiple SQL statements.

   .. method:: executemany(sql, parameters, /)

      For every item in *parameters*,
      repeatedly execute the :ref:`parameterized <sqlite3-placeholders>`
      SQL statement *sql*.

      Uses the same implicit transaction handling as :meth:`~Cursor.execute`.

      :param str sql:
         A single SQL :abbr:`DML (Data Manipulation Language)` statement.

      :param parameters:

>          A :term:`sequence` or :term:`iterator` of parameters

Reviewer: ```suggestion
         An :term:`iterable` or :term:`iterator` of parameters
```

>          An :term:`iterable` or :term:`iterator` of parameters

Reviewer: ```suggestion
         An :term:`iterable` of parameters
```

         to bind with the placeholders in *sql*.
         See :ref:`sqlite3-placeholders`.

>       :type parameters: :term:`iterable` | :term:`iterator`

Reviewer: ```suggestion
      :type parameters: :term:`iterable`
```

>       :type parameters: :term:`sequence` | :term:`iterator`

Reviewer: In typeshed we just use `Iterable` for this parameter rather than `Sequence | Iterator` — you could consider applying that simplification here too:

https://github.com/python/typeshed/blob/39487a8bab933a8c29edef5cf399de5fa8bb94cd/stdlib/sqlite3/dbapi2.pyi#L331

Author: Makes sense.
